design-proposals/storage-class-migration/storage-class-migration.md

```diff
@@ -0,0 +1,276 @@
+
+ ## Non  Goals
+*  This features is not directly aimed for live migrating the VM using RWO
+   volumes. Hence, the API covers and refers only to the storage migration.
+
+ ## Definition of Users
+ This feature addresses for all KubeVirt users who want to move to another storage classes or PVCs without workload interuptions.
+
+## User Stories
+  1. A storage class of volumes used by a running VM is deprecated
+  2. A more performant storage class is now available in the cluster and the user wants start using it.
+  3. The storage device is filling up
+
+## Repos
+
+  - [kubevirt/kubevirt](https://github.com/kubevirt/kubevirt)
+
+# Design
+This proposal introduces a new CRD `StorageMigration`
+
+```golang
+type StorageMigration struct {
+        metav1.TypeMeta   `json:",inline"`
+        metav1.ObjectMeta `json:"metadata,omitempty"`
+        Spec StorageMigrationSpec `json:"spec"`
+        Status *StorageMigrationStatus `json:"status,omitempty"`
+}
+
+// StorageMigrationList is a list of StorageMigration resources
+type StorageMigrationList struct {
+        metav1.TypeMeta `json:",inline"`
+        metav1.ListMeta `json:"metadata"`
+        Items []StorageMigration `json:"items"`
+}
+
+type MigratedVolume struct {
+        SourcePvc      string `json:"sourcePvc,omitempty" valid:"required"`
+        DestinationPvc string `json:"destinationPvc,omitempty" valid:"required"`
+}
+
+type MigrationStorageClass struct {
+        SourceStorageClass      string `json:"sourceStorageClass,omitempty" valid:"required"`
+        DestinationStorageClass string `json:"destinationStorageClass,omitempty" valid:"required"`
+        // PrefixDestinationVolume is the prefix used by KubeVirt to generate
+        // the destination PVC. The claim name will contain the prefix plus a
+        // random string.
+        PrefixDestinationVolume string `json:"PrefixDestinationVolume,omitempty" valid:"required"`
+}
+
+type ReclaimPolicySourcePvc string
+
+const (
+        DeleteReclaimPolicySourcePvc ReclaimPolicySourcePvc = "Delete"
+)
+
+// StorageMigrationSpec is the spec for a StorageMigration resource
+type StorageMigrationSpec struct {
+        // The name of the VMI to perform the migration on. VMI must exist in the storage migration objects
+        // namespace
+        VMIName string `json:"vmiName,omitempty" valid:"required"`
+        // MigrationVolumes is a list of volumes to migrate
+        MigrationVolumes []MigrationVolume `json:"migrationVolumes,omitempty"`
+        // MigrationStorageClass contains the information for migrating the
+        // volumes of the source storage class to the destination storage class
+        MigrationStorageClass *MigrationStorageClass `json:"migrationStorageClass,omitempty"`
+        // ReclaimPolicySourcePvc describes how the source volumes will be
+        // treated after a successful migration
+        ReclaimPolicySourcePvc ReclaimPolicySourcePvc `json:"reclaimPolicySourcePvc,omitempty"`
```

Comment:
What about having an optional reclaim policy also for each migration volume and storage class? Thus allowing a more precise tuning. E.g. if a volume or class does not have it set then the value from here can be used.

Reply:
Yes, it makes the logic more complex but we can of course do it. However, I would do for each storage class but not for each migrate volume as I think it becomes too complex for the user to specify

Reply:
I moved the policy to the `MigrationStorageClass` and `MigrationVolume` as you suggested